### PR TITLE
Add unit test support for Cosmos Client types

### PR DESCRIPTION
**Pull request by j82w**
To enable users to unit test the client all the existing classes were refactored to append Core to the end of class names, and made internal. A new abstract class was created and only exposes the public members. This will allow users to mock the abstract class.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/RequestOptions/CosmosRequestOptions.cs [hunk 1]
```diff
@@ -23,8 +23,8 @@ public class CosmosRequestOptions
         /// The condition (ETag) associated with the request.
         /// </value>
         /// <remarks>
-        /// Most commonly used with the Delete* and Replace* methods of <see cref="CosmosItems"/> such as <see cref="CosmosItems.ReplaceItemAsync{T}(object, string, T, CosmosItemRequestOptions, System.Threading.CancellationToken)"/>
-        /// but can be used with other methods like <see cref="CosmosItems.ReadItemAsync{T}(object, string, CosmosItemRequestOptions, System.Threading.CancellationToken)"/> for caching scenarios.
+        /// Most commonly used with the Delete* and Replace* methods of <see cref="CosmosItems"/> such as <see cref="CosmosItemsCore.ReplaceItemAsync{T}(object, string, T, CosmosItemRequestOptions, System.Threading.CancellationToken)"/>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/RequestOptions/CosmosRequestOptions.cs]: Docs should not use XXCore types. Only pubic abstract types only.
Shouldn't build fail for this?

**Comment by kirankumarkolli** [comment]
Thanks Jake.

File: Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs [hunk 2]
```diff
@@ -13,65 +14,65 @@ namespace Microsoft.Azure.Cosmos
...
 
         /// <summary>
-        /// The Cosmos resource URI
+        /// Initialize the common properties
         /// </summary>
-        internal Uri LinkUri { get; }
+        /// <param name="client">The cosmos client</param>
+        /// <param name="parentLink">The parent link URI (/dbs/mydbId)</param>
+        /// <param name="uriPathSegment">The URI path segment</param>
+        protected void Initialize(
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs]: nit: explicit access modifier (public/internal/private)?

File: Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs [hunk 3]
```diff
@@ -13,65 +14,65 @@ namespace Microsoft.Azure.Cosmos
...
+        }
 
         /// <summary>
         /// Generates the URI link for the resource
         /// </summary>
         /// <param name="parentLink">The parent link URI (/dbs/mydbId) </param>
+        /// <param name="uriPathSegment">The URI path segment</param>
         /// <returns>A resource link in the format of {parentLink}/this.UriPathSegment/this.Name with this.Name being a Uri escaped version</returns>
-        private string GetLinkString(string parentLink)
+        protected Uri GetLink(string parentLink, string uriPathSegment)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs]: nit: explicit access modifier (public/internal/private)?

File: Microsoft.Azure.Cosmos/src/Resource/Database/CosmosDatabasesCore.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    /// </code>
+    /// </example>
+    /// <example>
+    /// This example shows how to access an existing database. This does not do a network call or verify that the database exists in Cosmos.
+    /// <code language="c#">
+    /// <![CDATA[
+    /// CosmosDatabase database = this.cosmosClient.Databases["MyDatabaseId"];
+    /// ]]>
+    /// </code>
+    /// </example>
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Database/CosmosDatabasesCore.cs]: There is no real need to maintain docs with examples for internal implementations, this is more geared towards the public abstract classes, right?

File: Microsoft.Azure.Cosmos/src/Resource/Database/CosmosDatabasesCore.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    internal class CosmosDatabasesCore : CosmosDatabases
+    {
+        private readonly CosmosClient client;
+        private readonly ConcurrentDictionary<string, CosmosDatabase> databasesCache;
+
+        protected internal CosmosDatabasesCore(CosmosClient client)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Database/CosmosDatabasesCore.cs]: nit: explicit access modifier

File: Microsoft.Azure.Cosmos/src/Resource/Trigger/CosmosTrigger.cs [hunk 6]
```diff
@@ -30,13 +30,15 @@ internal class CosmosTrigger : CosmosIdentifier
         protected internal CosmosTrigger(
             CosmosContainer container,
             string triggerId)
-            : base(container.Client,
-                container.Link,
-                triggerId)
         {
+            this.Id = triggerId;
+            base.Initialize(
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Trigger/CosmosTrigger.cs]: I wonder if there is a way to enforce the call to Initialize? The previous version (through inheritance) made sure that the base constructor was called. With Initialize, it is possible that the child does not call it accidentally, right?

File: Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs [hunk 7]
```diff
@@ -13,65 +14,65 @@ namespace Microsoft.Azure.Cosmos
...
         /// The Id of the cosmos resource
         /// </summary>
-        public virtual string Id { get; }
+        public abstract string Id { get; }
 
         /// <summary>
         /// The Cosmos client that is used for the request
         /// </summary>
-        internal CosmosClient Client { get; }
+        internal CosmosClient Client { get; private set; }
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs]: From a mocking point of view, this property would not be accessible to the mocked version, right? What if a mocked CosmosIdentifier is sent to a class that expects the Client to be set?

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs]: Its internal and mock can't set it. Also most UT are for outward contract then E2E integration testing.

File: Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/access-modifiers

File: Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosIdentifier.cs]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/access-modifiers

File: Microsoft.Azure.Cosmos/src/Resource/Trigger/CosmosTrigger.cs (hunk 6 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Trigger/CosmosTrigger.cs]: There is a work item to remove the base class. There are tests that verify this is set and works properly.

**Pull request merged**